coderd/workspaceapps/proxy.go

```diff
@@ -0,0 +1,594 @@
+			// configured.
+			if s.AppHostname == "" || s.AppHostnameRegex == nil {
+				next.ServeHTTP(rw, r)
+				return
+			}
+
+			// Step 2: Get the request Host.
+			host := httpapi.RequestHost(r)
+			if host == "" {
+				if r.URL.Path == "/derp" {
+					// The /derp endpoint is used by wireguard clients to tunnel
+					// through coderd. For some reason these requests don't set
+					// a Host header properly sometimes in tests (no idea how),
+					// which causes this path to get hit.
+					next.ServeHTTP(rw, r)
+					return
+				}
+
+				httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+					Message: "Could not determine request Host.",
+				})
+				return
+			}
+
+			// Steps 3-6: Parse application from subdomain.
+			app, ok := s.parseWorkspaceApplicationHostname(rw, r, next, host)
+			if !ok {
+				return
+			}
+
+			// If the request has the special query param then we need to set a
+			// cookie and strip that query parameter.
+			if encryptedAPIKey := r.URL.Query().Get(SubdomainProxyAPIKeyParam); encryptedAPIKey != "" {
+				// Exchange the encoded API key for a real one.
+				token, err := s.AppSigningKey.DecryptAPIKey(encryptedAPIKey)
+				if err != nil {
+					s.Logger.Debug(ctx, "could not decrypt API key", slog.Error(err))
+					site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+						Status:      http.StatusBadRequest,
+						Title:       "Bad Request",
+						Description: "Could not decrypt API key. Please remove the query parameter and try again.",
+						// Retry is disabled because the user needs to remove
+						// the query parameter before they try again.
+						RetryEnabled: false,
+						DashboardURL: s.AccessURL.String(),
+					})
+					return
+				}
+
+				s.setWorkspaceAppCookie(rw, r, token)
+
+				// Strip the query parameter.
+				path := r.URL.Path
+				if path == "" {
+					path = "/"
+				}
+				q := r.URL.Query()
+				q.Del(SubdomainProxyAPIKeyParam)
+				rawQuery := q.Encode()
+				if rawQuery != "" {
+					path += "?" + q.Encode()
+				}
+
+				http.Redirect(rw, r, path, http.StatusTemporaryRedirect)
+				return
+			}
+
+			token, ok := ResolveRequest(s.Logger, s.AccessURL, s.TokenProvider, rw, r, Request{
+				AccessMethod:      AccessMethodSubdomain,
+				BasePath:          "/",
+				UsernameOrID:      app.Username,
+				WorkspaceNameOrID: app.WorkspaceName,
+				AgentNameOrID:     app.AgentName,
+				AppSlugOrPort:     app.AppSlugOrPort,
+			})
+			if !ok {
+				return
+			}
+
+			// Use the passed in app middlewares before passing to the proxy
+			// app.
+			mws := chi.Middlewares(middlewares)
+			mws.Handler(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+				s.proxyWorkspaceApp(rw, r, *token, r.URL.Path)
+			})).ServeHTTP(rw, r.WithContext(ctx))
+		})
+	}
+}
+
+func (s *WorkspaceAppServer) parseWorkspaceApplicationHostname(rw http.ResponseWriter, r *http.Request, next http.Handler, host string) (httpapi.ApplicationURL, bool) {
```

Comment:
I might remove naming duplication from these handler names too.

It will help focus on the idea that all routes in here are focused to `workspaceApps`, instead of just some.